game-core/src/main/java/games/strategy/triplea/delegate/dice/roller/RollNDice.java

```diff
@@ -0,0 +1,37 @@
+package games.strategy.triplea.delegate.dice.roller;
+
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.engine.random.IRandomStats;
+import games.strategy.triplea.delegate.DiceRoll;
+import games.strategy.triplea.delegate.Die;
+import java.util.ArrayList;
+import java.util.List;
+import lombok.experimental.UtilityClass;
+
+/**
+ * Roll n-sided dice.
+ *
+ * <p>0 based, add 1 to get actual die roll
+ */
+@UtilityClass
+public class RollNDice {
```

Comment:
RollNDice could be confused with "rollinDice" and "roll and dice". It is also confusing if we are rolling a N-sided die or are rolling n-dice.

This class looks like a dice-roller. I would opt to name it that. Looking at the one method in this class, I notice the API is not the prettiest. Of note, we have two int parameters side by side that could be swapped or would be hard to distinguish.

One fix would be to convert the parameters of the method to private variables and then use a builder to construct and have a no-arg 'rollDice' method. That should clean up the usage. Though, a lot of the parameters are simply to feed the call to "bridge.getRandom(sides, rollCount, playerRolling, diceType, annotation);". I think this method might be trying to do too much an dperhaps should take a `random[]` as a parameter and then it would have far fewer methods. Probably even better would be to have `DiceRoll` to have a constructor that accepts the `random[]` computation and it can then do most of the rest of the logic. 

TL;DR, this method looks like it should instead be a constructor on DiceRoll.

Reply:
This class is a builder of `DiceRoll`.  You shouldn't create your own `DiceRoll` but instead call this method if you need to roll `N` dice.  This also creates special `IGNORED` dice.  You don't want to use this for battle situations.  This is currently used by war bonds but I want to expand it to the other places that call `bridge.getRandom` so that we have a unified dice logic.

I thought of calling this `RollIgnoredDice` but that would be really strange.  It was originally called `rollNDice` and I think it fits.

Reply:
### RE: rollNDice

When I first read 'rollNDice', my first thought it _was_ 'rollAndDice' and it was similar to 'StengthAndRolls'

I'd say then the original name was clever, too short and required more than naive context to understand (which means it takes longer to understand the code, the more context you have to acquire, because context acquisition takes time, ergo the more context needed implies more time is needed to read and understand code).

TBH I'm still not clear if `n` is the number of dice or the number of dice sides. I think the latter, but maybe it's even both??

So please have some thoughts about the naming for `rollNDice` and consider spelling it out to make it easier on a reader.


### Re: Factory methods

Okay, it's good to know this is supposed to be a factory method. I think the cohesion heuristic is potentially low as we have a factory class, with a factory method, that are in one package, meanwhile `DiceRoll` is in: `package games.strategy.triplea.delegate;`

My concern about the 3 factory classes is that they are each relatively shallow, it's not offhand clear they are factories either explicitly or from the javadocs, and it's potentially not clear when to use and not the other. 

To alleviate these concerns I can suggest we combine the factory methods into one class and then document when you would use one each one. Seeing the series of methods in one class file I think would draw context between when each one would be used. I think ideally the factory class would be 'close' by to the 'DiceRoll' class, though I can understand not having them co-located just yet.

WDYT?

Reply:
@DanVanAtta , I've renamed `RollNDice` to `RollNSidedDiceXTimes`.  I hope that is more specific.

As for the changes to the factory methods, I want to move `DiceRoll` but I can't because of saved game issue.  I've already added `@ChangeOnNextMajorRelease` to indicate that `DiceRoll` should be moved.  Once it is moved, it should make more sense.  I understand your desire to merge them into one class with 3 static methods but I hope to change them to `@Builder`s and I plan to try and get more of the `getRandom` usages added as new factory classes.  So I'd rather leave them separate for now.

Reply:
I also thought the javadocs explain when you would use one or the other.  I'm not sure what to add to them to make them clearer.

Reply:
`RollNSidedDiceXTimes` is wordy but certainly more clear!

My thinking was with the factory classes merged into one, the existing javadocs would contrast against one another more readily and make it obvious that the factory methods were related. That is an off-the-cuff suggestion, I'll take a deeper look soon in case there is a better suggestion that comes to mind or if my initial impression was just off-base.